### PR TITLE
DHT Sensor support

### PR DESCRIPTION
**Pull request by ghost**
Adding the option to read the local ambient temp and humidity from a DHT
sensor.

### CODE REVIEW

File: Marlin/Configuration.h [hunk 1]
```diff
@@ -1460,4 +1460,26 @@
...
+ *         http://www.github.com/markruys/arduino-DHT
+ *         *******************************************************
+ * */
+#define DHT_ENABLE
+#if ENABLED(DHT_ENABLE)
+    #define DHT_TYPE 22     //DHT type 11, 21, or 22
+    #define DHT_PIN A3      //Set pin for the DHT sensor. Default = A3 (AUX1 connector pin3 (outside row))
+
+#endif //DHT_ENABLE
+
```
Comment (Blue-Marlin) [inline Marlin/Configuration.h]: There are much more Configuration.h-files in the `example_configurations` directory. All of them have to be modified.

Maybe `Configuration_adv.h` would be the better place for this feature.

File: Marlin/language_en.h [hunk 2]
```diff
@@ -462,6 +462,9 @@
 #ifndef MSG_HEATING
   #define MSG_HEATING                         _UxGT("Heating...")
 #endif
+#ifndef MSG_DHT_WARMING
+  #define MSG_DHT_WARMING                     _UxGT("Sensor Warming...")
+#endif
```
Comment (Blue-Marlin) [inline Marlin/language_en.h]: Seems to be unused.

File: Marlin/ultralcd.cpp [hunk 3]
```diff
@@ -2389,6 +2396,34 @@ void kill_screen(const char* lcd_msg) {
...
+          DHT_temperature = (dht.getTemperature());
+          DHT_humidity = (dht.getHumidity());
+          DHT_done = true;
+        }
+        START_SCREEN();
+          STATIC_ITEM("Ambient");                                                    //Ambient
+          STATIC_ITEM(MSG_INFO_DHT_MENU);                                            //Temp & Humidity
+          STATIC_ITEM("    Temp (C): ", false, true, itostr3left(DHT_temperature));  //Temp (C): 14
+          STATIC_ITEM("    Humidity: ", false, true, itostr3left(DHT_humidity));     //Humidity: 27
+        END_SCREEN();
```
Comment (Blue-Marlin) [inline Marlin/ultralcd.cpp]: While this strings should be translatable.

File: Marlin/ultralcd.cpp [hunk 4]
```diff
@@ -2389,6 +2396,34 @@ void kill_screen(const char* lcd_msg) {
...
+    #if ENABLED(DHT_ENABLE)
+
+      void lcd_info_dht_menu(){
+        if (lcd_clicked) { return lcd_goto_previous_menu(); }
+        if (DHT_done == false)
+        {
+          DHT dht;
+          dht.setup(DHT_PIN, DHT_TYPE);
+          safe_delay (1800);
+          DHT_temperature = (dht.getTemperature());
```
Comment (Blue-Marlin) [inline Marlin/ultralcd.cpp]: 'safe_delay' is a good solution here - but the ' ' before the '(' is ugly.
```diff
-          DHT_temperature = (dht.getTemperature());
+          DHT_temperature = dht.getTemperature();
```

File: Marlin/language_en.h (hunk 2 above)
Comment (ghost) [reply Marlin/language_en.h]: Yes...I wanted to display "Sensor Warming..." during the delay. I am still trying to nail down how to create that screen.

File: Marlin/Configuration.h (hunk 1 above)
Comment (ghost) [reply Marlin/Configuration.h]: PITA, but done.

File: Marlin/ultralcd.cpp (hunk 4 above)
Comment (ghost) [reply Marlin/ultralcd.cpp]: Corrected for your viewing pleasure. :)

File: Marlin/ultralcd.cpp (hunk 3 above)
Comment (ghost) [reply Marlin/ultralcd.cpp]: I thought about that. Done.

File: Marlin/SanityCheck.h [hunk 5]
```diff
@@ -261,6 +261,15 @@
 #endif
 
 /**
+ * DHT Sensor option
+ */
+#if ENABLED(DHT_ENABLE)
+  #if DISABLED(ULTIPANEL)
+    #error "DHT Sensor option requires an LCD controller."
+  #endif
+#endif
```
Comment (akaJes) [inline Marlin/SanityCheck.h]: also it need LCD_INFO_MENU enabled

Comment (Tannoo) [inline Marlin/SanityCheck.h]: How about this instead?

`#define DHT_ENABLE`
`#if ENABLED(DHT_ENABLE)`
    `#ifndef LCD_INFO_MENU`
      `#define LCD_INFO_MENU //This must be defined.`
    `#endif`
    `#define DHT_TYPE 22     //DHT type 11, 21, or 22`
    `#define DHT_PIN A3      //Set pin for the DHT sensor.`
`#endif //DHT_ENABLE`

Comment (Tannoo) [inline Marlin/SanityCheck.h]: In `configuration_adv.h`.

Comment (ghost) [reply Marlin/SanityCheck.h]: You are correct. Done.

Comment (ghost) [reply Marlin/SanityCheck.h]: Should I move this to sanity check instead of automaticly enabling the `lcd_info_menu`?

Comment (Roxy-3D) [reply Marlin/SanityCheck.h]: Yes...   Sanity.cpp would be the best place for this check.   But I presume you would be putting a:
```cpp
#if ENABLED(DHT_ENABLE)
```
in front of that check, right?

Comment (ghost) [reply Marlin/SanityCheck.h]: Of course. :)

Comment (ghost) [reply Marlin/SanityCheck.h]: There is no `Sanity.cpp`. There is a `SanityCheck.h`.

**Comment by thinkyhead** [comment]
Looks like this PR got lost in the shuffle while I was being homeless. Are you planning to re-submit?

**Pull request closed without merging**